crates/miden-objects/src/account/storage/map/witness.rs

```diff
@@ -1,50 +1,112 @@
+/// valid state.
 #[derive(Debug, Clone, PartialEq, Eq)]
-pub struct StorageMapWitness(SmtProof);
+pub struct StorageMapWitness {
+    proof: SmtProof,
+    map: BTreeMap<Word, Word>,
+}
 
 impl StorageMapWitness {
     // CONSTRUCTORS
     // --------------------------------------------------------------------------------------------
 
-    /// Creates a new [`StorageMapWitness`] from an SMT proof.
-    pub fn new(smt_proof: SmtProof) -> Self {
-        Self(smt_proof)
+    /// Creates a new [`StorageMapWitness`] from an SMT proof and a provided set of map keys.
+    ///
+    /// # Errors
+    ///
+    /// Returns an error if:
+    /// - Any of the map keys is not contained in the proof.
+    pub fn new(
+        proof: SmtProof,
+        map_keys: impl IntoIterator<Item = Word>,
+    ) -> Result<Self, StorageMapError> {
+        let mut map = BTreeMap::new();
+
+        for map_key in map_keys.into_iter() {
+            let hashed_map_key = StorageMap::hash_key(map_key);
+            let value =
+                proof.get(&hashed_map_key).ok_or(StorageMapError::MissingKey { map_key })?;
+            map.insert(map_key, value);
+        }
+
+        Ok(Self { proof, map })
+    }
+
+    /// Creates a new [`StorageMapWitness`] from an SMT proof and a set of key value pairs.
+    ///
+    /// # Warning
+    ///
+    /// This does not validate any of the guarantees of this type. See the type-level docs for more
+    /// details.
+    pub fn new_unchecked(
+        proof: SmtProof,
+        key_values: impl IntoIterator<Item = (Word, Word)>,
+    ) -> Self {
+        Self {
+            proof,
+            map: key_values.into_iter().collect(),
+        }
     }
 
     // PUBLIC ACCESSORS
     // --------------------------------------------------------------------------------------------
 
-    /// Searches for a value in the witness with the given `map_key`.
-    pub fn find(&self, hashed_map_key: Word) -> Option<Word> {
-        self.entries()
+    /// Returns the underlying [`SmtProof`].
+    pub fn as_proof(&self) -> &SmtProof {
+        &self.proof
+    }
+
+    /// Searches for a value in the witness with the given unhashed `map_key`.
+    pub fn find(&self, map_key: Word) -> Option<Word> {
+        let hashed_map_key = StorageMap::hash_key(map_key);
+        self.hashed_entries()
             .find_map(|(key, value)| if *key == hashed_map_key { Some(*value) } else { None })
     }
 
     /// Returns an iterator over the key-value pairs in this witness.
     ///
-    /// Note that the returned key is the hashed map key.
+    /// Note that the returned key is the original map key.
     pub fn entries(&self) -> impl Iterator<Item = (&Word, &Word)> {
+        self.map.iter()
+    }
+
+    /// Returns an iterator over the key-value pairs in this witness.
+    ///
+    /// Note that the returned key is the hashed map key.
+    pub fn hashed_entries(&self) -> impl Iterator<Item = (&Word, &Word)> {
```

Comment:
Does this method need to be public?

Reply:
It was used in `find`, but now that we have the `entries` map, I renamed `find` to `get` and lookup the key in the `BTreeMap` directly and removed `hashed_entries`.